roachtest: tpchbench/tpchVec/nodes=3/cpu=4/sf=1 failed

SHA: https://github.com/cockroachdb/cockroach/commits/8c6fdc64908a13291e4ddc5d233bbbaa379e71a2

Parameters:

To repro, try:

```
# Don't forget to check out a clean suitable branch and experiment with the
# stress invocation until the desired results present themselves. For example,
# using stress instead of stressrace and passing the '-p' stressflag which
# controls concurrency.
./scripts/gceworker.sh start && ./scripts/gceworker.sh mosh
cd ~/go/src/github.com/cockroachdb/cockroach && \
stdbuf -oL -eL \
make stressrace TESTS=tpchbench/tpchVec/nodes=3/cpu=4/sf=1 PKG=roachtest TESTTIMEOUT=5m STRESSFLAGS='-maxtime 20m -timeout 10m' 2>&1 | tee /tmp/stress.log
```

Failed test: https://teamcity.cockroachdb.com/viewLog.html?buildId=1378458&tab=buildLog

```
The test failed on branch=master, cloud=gce:
test artifacts and logs in: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/artifacts/20190709-1378458/tpchbench/tpchVec/nodes=3/cpu=4/sf=1/run_1
	tpchbench.go:119,cluster.go:2067,errgroup.go:57: /home/agent/work/.go/src/github.com/cockroachdb/cockroach/bin/roachprod run teamcity-1562652995-02-n4cpu4:4 -- ./workload run querybench --db=tpch --concurrency=1 --query-file=tpchVec --num-runs=3 --max-ops=27 --vectorized=true {pgurl:1-3} --histograms=logs/stats.json --histograms-max-latency=8m20s returned:
		stderr:
		
		stdout:
		      0            1.0            0.0   1476.4   1476.4   1476.4   1476.4  8: SELECT ps_partkey, sum(ps_supplycost * ps_availqty::float) AS value FROM partsupp, supplier, nation WHERE ps_suppkey = s_suppkey AND s_nationkey = n_nationkey AND n_name = 'GERMANY' GROUP BY ps_partkey HAVING sum(ps_supplycost * ps_availqty::float) > ( SELECT sum(ps_supplycost * ps_availqty::float) * 0.0001 FROM partsupp, supplier, nation WHERE ps_suppkey = s_suppkey AND s_nationkey = n_nationkey AND n_name = 'GERMANY') ORDER BY value DESC
		  16m39s        0            0.0            0.0      0.0      0.0      0.0      0.0  9: SELECT sum(l_extendedprice) / 7.0 AS avg_yearly FROM lineitem, part WHERE p_partkey = l_partkey AND p_brand = 'Brand#23' AND p_container = 'MED BOX' AND l_quantity < ( SELECT 0.2 * avg(l_quantity) FROM lineitem WHERE l_partkey = p_partkey)
		Error: pq: mismatched batch width and schema length: 1 != 2
		Error:  ssh verbose log retained in /root/.roachprod/debug/ssh_35.243.149.128_2019-07-09T07:07:28Z: exit status 1
		: exit status 1
	cluster.go:2088,tpchbench.go:123,tpchbench.go:244,test_runner.go:670: Goexit() was called

```

```
panic: send on closed channel

goroutine 4284 [running]:
panic(0x3835080, 0x4434670)
	/usr/local/go/src/runtime/panic.go:565 +0x2c5 fp=0xc00c1ef810 sp=0xc00c1ef780 pc=0x785f75
runtime.chansend(0xc00c55e0c0, 0xc00c1ef980, 0x38b1801, 0x2508aa1, 0xc01c548830)
	/usr/local/go/src/runtime/chan.go:187 +0x5f5 fp=0xc00c1ef890 sp=0xc00c1ef810 pc=0x75dd15
runtime.chansend1(0xc00c55e0c0, 0xc00c1ef980)
	/usr/local/go/src/runtime/chan.go:127 +0x35 fp=0xc00c1ef8c8 sp=0xc00c1ef890 pc=0x75d715
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Inbox).init(0xc0003e9ce0, 0x44ec800, 0xc00357edb0, 0x7645d9, 0xc00c1efa78)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/inbox.go:145 +0x2e1 fp=0xc00c1efa38 sp=0xc00c1ef8c8 pc=0x2508aa1
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Inbox).maybeInit(0xc0003e9ce0, 0x44ec800, 0xc00357edb0, 0x40, 0x3f)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/inbox.go:126 +0x57 fp=0xc00c1efa70 sp=0xc00c1efa38 pc=0x2508787
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Inbox).DrainMeta(0xc0003e9ce0, 0x44ec800, 0xc00357edb0, 0x0, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/inbox.go:282 +0xda fp=0xc00c1efc00 sp=0xc00c1efa70 pc=0x2509b6a
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).sendMetadata(0xc0005b0d20, 0x44ec800, 0xc00357edb0, 0x7fc354a62500, 0xc006e9f020, 0x447a540, 0xc000349000, 0x0, 0x0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:238 +0xd2 fp=0xc00c1efdf8 sp=0xc00c1efc00 pc=0x250b352
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).runWithStream(0xc0005b0d20, 0x44ec800, 0xc00357edb0, 0x7fc354a62500, 0xc006e9f020, 0xc01c548370)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:279 +0x182 fp=0xc00c1efe70 sp=0xc00c1efdf8 pc=0x250b8a2
github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc.(*Outbox).Run(0xc0005b0d20, 0x44ec740, 0xc01c312040, 0xc000bdc7e0, 0xc60d56a00000001, 0x1c0d05a31b48e094, 0xc0bd69d464, 0x41, 0xc01c548370)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/exec/colrpc/outbox.go:147 +0x54c fp=0xc00c1eff40 sp=0xc00c1efe70 pc=0x250aacc
github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1.1(0xc0005b0d20, 0x44ec740, 0xc01c312040, 0xc000544000, 0xc00372f120, 0xc01c548370, 0xc0005441a0)
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:851 +0x8c fp=0xc00c1effa8 sp=0xc00c1eff40 pc=0x258b6bc
runtime.goexit()
	/usr/local/go/src/runtime/asm_amd64.s:1337 +0x1 fp=0xc00c1effb0 sp=0xc00c1effa8 pc=0x7b5d11
created by github.com/cockroachdb/cockroach/pkg/sql/distsqlrun.(*Flow).setupVectorizedRemoteOutputStream.func1
	/go/src/github.com/cockroachdb/cockroach/pkg/sql/distsqlrun/column_exec_setup.go:850 +0x87
```

--- a/pkg/sql/exec/testutils.go
+++ b/pkg/sql/exec/testutils.go
@@ -100,3 +100,17 @@ func (s *RepeatableBatchSource) ResetBatchesToReturn(b int) {
 	s.batchesToReturn = b
 	s.batchesReturned = 0
 }
+
+// CallbackOperator is a testing utility struct that delegates Next calls to a
+// callback provided by the user.
+type CallbackOperator struct {
+	NextCb func(ctx context.Context) coldata.Batch
+}
+
+// Init is part of the Operator interface.
+func (o *CallbackOperator) Init() {}
+
+// Next is part of the Operator interface.
+func (o *CallbackOperator) Next(ctx context.Context) coldata.Batch {
+	return o.NextCb(ctx)
+}

--- a/pkg/sql/exec/unorderedsynchronizer.go
+++ b/pkg/sql/exec/unorderedsynchronizer.go
@@ -49,13 +49,26 @@ type UnorderedSynchronizer struct {
 	initialized bool
 	done        bool
 	zeroBatch   coldata.Batch
-	wg          *sync.WaitGroup
-	cancelFn    context.CancelFunc
-	batchCh     chan *unorderedSynchronizerMsg
-	errCh       chan error
+	// externalWaitGroup refers to the WaitGroup passed in externally. Since the
+	// UnorderedSynchronizer spawns goroutines, this allows callers to wait for
+	// the completion of these goroutines.
+	externalWaitGroup *sync.WaitGroup
+	// internalWaitGroup refers to the WaitGroup internally managed by the
+	// UnorderedSynchronizer. This will only ever be incremented by the
+	// UnorderedSynchronizer and decremented by the input goroutines. This allows
+	// the UnorderedSynchronizer to wait only on internal goroutines.
+	internalWaitGroup *sync.WaitGroup
+	cancelFn          context.CancelFunc
+	batchCh           chan *unorderedSynchronizerMsg
+	errCh             chan error
 }
 
-// NewUnorderedSynchronizer creates a new UnorderedSynchronizer.
+// NewUnorderedSynchronizer creates a new UnorderedSynchronizer. On the first
+// call to Next, len(inputs) goroutines are spawned to read each input
+// asynchronously (to not be limited by a slow input). These will increment
+// the passed-in WaitGroup and decrement when done. It is also guaranteed that
+// these spawned goroutines will have completed on any error or zero-length
+// batch received from Next.
 func NewUnorderedSynchronizer(
 	inputs []Operator, typs []types.T, wg *sync.WaitGroup,
 ) *UnorderedSynchronizer {
@@ -68,11 +81,12 @@ func NewUnorderedSynchronizer(
 	zeroBatch := coldata.NewMemBatchWithSize(typs, 0)
 	zeroBatch.SetLength(0)
 	return &UnorderedSynchronizer{
-		inputs:        inputs,
-		readNextBatch: readNextBatch,
-		zeroBatch:     zeroBatch,
-		wg:            wg,
-		batchCh:       make(chan *unorderedSynchronizerMsg, len(inputs)),
+		inputs:            inputs,
+		readNextBatch:     readNextBatch,
+		zeroBatch:         zeroBatch,
+		externalWaitGroup: wg,
+		internalWaitGroup: &sync.WaitGroup{},
+		batchCh:           make(chan *unorderedSynchronizerMsg, len(inputs)),
 		// errCh is buffered so that writers do not block. If errCh is full, the
 		// input goroutines will not push an error and exit immediately, given that
 		// the Next goroutine will read an error and panic anyway.
@@ -98,7 +112,8 @@ func (s *UnorderedSynchronizer) Init() {
 func (s *UnorderedSynchronizer) init(ctx context.Context) {
 	ctx, s.cancelFn = contextutil.WithCancel(ctx)
 	for i, input := range s.inputs {
-		s.wg.Add(1)
+		s.externalWaitGroup.Add(1)
+		s.internalWaitGroup.Add(1)
 		// TODO(asubiotto): Most inputs are Inboxes, and these have handler
 		// goroutines just sitting around waiting for cancellation. I wonder if we
 		// could reuse those goroutines to push batches to batchCh directly.
@@ -107,7 +122,8 @@ func (s *UnorderedSynchronizer) init(ctx context.Context) {
 				if int(atomic.AddUint32(&s.numFinishedInputs, 1)) == len(s.inputs) {
 					close(s.batchCh)
 				}
-				s.wg.Done()
+				s.internalWaitGroup.Done()
+				s.externalWaitGroup.Done()
 			}()
 			msg := &unorderedSynchronizerMsg{
 				inputIdx: inputIdx,
@@ -176,12 +192,14 @@ func (s *UnorderedSynchronizer) Next(ctx context.Context) coldata.Batch {
 			// If we got an error from one of our inputs, cancel all inputs and
 			// propagate this error through a panic.
 			s.cancelFn()
+			s.internalWaitGroup.Wait()
 			panic(err)
 		}
 	case msg := <-s.batchCh:
 		if msg == nil {
-			// All inputs have exited, check if this was a graceful termination or
-			// not.
+			// All inputs have exited, double check that this is indeed the case.
+			s.internalWaitGroup.Wait()
+			// Check if this was a graceful termination or not.
 			select {
 			case err := <-s.errCh:
 				if err != nil {

--- a/pkg/sql/exec/unorderedsynchronizer_test.go
+++ b/pkg/sql/exec/unorderedsynchronizer_test.go
@@ -13,6 +13,7 @@ package exec
 import (
 	"context"
 	"sync"
+	"sync/atomic"
 	"testing"
 	"time"
 
@@ -87,6 +88,33 @@ func TestUnorderedSynchronizer(t *testing.T) {
 		require.Equal(t, numInputs*numBatches, batchesReturned)
 	}
 	wg.Wait()
+}
+
+func TestUnorderedSynchronizerNoLeaksOnError(t *testing.T) {
+	defer leaktest.AfterTest(t)()
+
+	const expectedErr = "first input error"
+
+	inputs := make([]Operator, 6)
+	inputs[0] = &CallbackOperator{NextCb: func(context.Context) coldata.Batch { panic(expectedErr) }}
+	for i := 1; i < len(inputs); i++ {
+		inputs[i] = &CallbackOperator{
+			NextCb: func(ctx context.Context) coldata.Batch {
+				<-ctx.Done()
+				panic(ctx.Err())
+			},
+		}
+	}
+
+	var (
+		ctx = context.Background()
+		wg  sync.WaitGroup
+	)
+	s := NewUnorderedSynchronizer(inputs, []types.T{types.Int64}, &wg)
+	err := CatchVectorizedRuntimeError(func() { _ = s.Next(ctx) })
+	// This is the crux of the test: assert that all inputs have finished.
+	require.Equal(t, len(inputs), int(atomic.LoadUint32(&s.numFinishedInputs)))
+	require.True(t, testutils.IsError(err, expectedErr), err)
 }
 
 func BenchmarkUnorderedSynchronizer(b *testing.B) {

